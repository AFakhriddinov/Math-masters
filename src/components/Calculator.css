.main {
  width: 600px;
  background-color: rgb(237, 225, 225);
  display: flex;
  flex-direction: column;
  margin: auto;
}

.operand-input {
  display: flex;
  flex-direction: column;
}

.operand-input input {
  border: none;
  outline: none;
  direction: rtl;
  color: white;
  height: 70px;
  background-color: gray;
  padding-right: 10px;
  font-size: 30px;
}

.buttons {
  display: flex;
  height: 500px;
}

.buttons .left-section {
  flex: 2;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
}

.buttons .left-section > * {
  padding: 10px 25px;
  border: 1px solid lightgray;
  font-size: 28px;
  color: black;
  font-weight: bold;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}

.buttons .left-section .zero {
  grid-column: 1 / span 2;
}

.buttons .right-section {
  flex: 1;
  display: grid;
  flex-direction: column;
}

.buttons .right-section > * {
  padding: 10px 25px;
  background-color: orange;
  border: 1px solid lightgray;
  font-size: 28px;
  color: black;
  font-weight: bold;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
}
