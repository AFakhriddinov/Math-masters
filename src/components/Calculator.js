/* eslint-disable react/prop-types */
import React, { useState } from 'react';
import '../App.css';
import calculate from './logic/calculate';

const Calculator = () => {
  // const [state, updateState] = useState({
  //   total: null,
  //   operation: null,
  //   next: null,
  // });

  // const calculation = (e) => {
  //   const newState = calculate(state, e.target.textContent);
  //   updateState(newState);
  // };
  const [state, SetState] = useState({});
  const calculation = (e) => {
    SetState(calculate(state, e.target.textContent));
  };
  const { next, total, operation } = state;
  const display = (total || '') + (operation || '') + (next || '') || '0';

  return (
    <div className="main-section">
      <h2>Lets do some Math</h2>
      <div className="main">
        {/* <Operand data={state} /> */}
        <div className="operand-input" data-testid="display">
          {display}
        </div>
        <div className="buttons">
          <div className="left-section">
            <button type="button" onClick={calculation} className="ac">
              AC
            </button>
            <button type="button" onClick={calculation} className="plusOrMinus">
              +/-
            </button>
            <button type="button" onClick={calculation} className="modelus">
              %
            </button>
            <button type="button" onClick={calculation} className="seven">
              7
            </button>
            <button type="button" onClick={calculation} className="eight">
              8
            </button>
            <button type="button" onClick={calculation} className="nine">
              9
            </button>
            <button type="button" onClick={calculation} className="four">
              4
            </button>
            <button type="button" onClick={calculation} className="five">
              5
            </button>
            <button type="button" onClick={calculation} className="six">
              6
            </button>
            <button type="button" onClick={calculation} className="1">
              1
            </button>
            <button type="button" onClick={calculation} className="two">
              2
            </button>
            <button type="button" onClick={calculation} className="three">
              3
            </button>
            <button type="button" onClick={calculation} className="zero">
              0
            </button>
            <button type="button" onClick={calculation} className="dot">
              .
            </button>
          </div>
          <div className="right-section">
            <button type="button" onClick={calculation} className="plus">
              /
            </button>
            <button type="button" onClick={calculation} className="plus">
              x
            </button>
            <button type="button" onClick={calculation} className="plus">
              -
            </button>
            <button type="button" onClick={calculation} className="plus">
              +
            </button>
            <button type="button" onClick={calculation} className="plus">
              =
            </button>
          </div>
        </div>
      </div>
    </div>
  );

  // function Operand(props) {
  //   const state = props;
  //   return (
  //     <div className="operand-input">
  //       <input
  //         data-testid="display"
  //         type="text"
  //         name="operand"
  //         value={state.data.total || state.data.operation || state.data.next}
  //       />
  //     </div>
  //   );
  // }
};
export default Calculator;
