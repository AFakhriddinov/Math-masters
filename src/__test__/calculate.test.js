import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Calculator from '../components/Calculator';
import calculate from '../components/logic/calculate';

describe('Test calculate function', () => {
  const obj = {
    total: null,
    next: null,
    operation: null,
  };
  it('Plus opetaion', () => {
    render(<Calculator />);
    const three = screen.getByText('3');
    const sum = screen.getByText('+');
    const two = screen.getByText('2');
    const equals = screen.getByText('=');
    const display = screen.getByTestId('display');
    fireEvent.click(three);
    fireEvent.click(sum);
    fireEvent.click(two);
    fireEvent.click(equals);
    expect(display.innerHTML).toBe('5');
  });

  it('Minus operation', () => {
    render(<Calculator />);
    const nine = screen.getByText('9');
    const sub = screen.getByText('-');
    const six = screen.getByText('6');
    const equals = screen.getByText('=');
    const display = screen.getByTestId('display');
    fireEvent.click(nine);
    fireEvent.click(sub);
    fireEvent.click(six);
    fireEvent.click(equals);
    expect(display.innerHTML).toBe('3');
  });

  it('Testing that Calculator renders correctly', async () => {
    const { asFragment } = render(<Calculator />);
    expect(asFragment()).toMatchSnapshot();
  });

  obj.next = '1';
  test('Pressing button +/- and next is 1 it will returns -1', () => {
    expect(calculate(obj, '+/-')).toEqual({
      total: null,
      next: '-1',
      operation: null,
    });
  });

  test('When next is 5 and press the . button, it will returns 5.', () => {
    const testOperation = { total: null, next: '5', operation: '.' };
    expect(calculate(testOperation, '.')).toEqual({
      total: null,
      next: '5.',
      operation: '.',
    });
  });

  test('User pressed an operation + after pressing = button', () => {
    const testOperation = { total: '10', next: null, operation: '+' };
    expect(calculate(testOperation, '+')).toEqual({
      total: '10',
      next: null,
      operation: '+',
    });
  });

  test('User pressed an operation + button and there is an existing operation x', () => {
    const testOperation = { total: '10', next: null, operation: 'x' };
    expect(calculate(testOperation, '+')).toEqual({
      total: '10',
      next: null,
      operation: '+',
    });
  });
});
