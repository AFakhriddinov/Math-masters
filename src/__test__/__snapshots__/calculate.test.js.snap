// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test calculate function Testing that Calculator renders correctly 1`] = `
<DocumentFragment>
  <div
    class="main-section"
  >
    <h2>
      Lets do some Math
    </h2>
    <div
      class="main"
    >
      <div
        class="operand-input"
        data-testid="display"
      >
        0
      </div>
      <div
        class="buttons"
      >
        <div
          class="left-section"
        >
          <button
            class="ac"
            type="button"
          >
            AC
          </button>
          <button
            class="plusOrMinus"
            type="button"
          >
            +/-
          </button>
          <button
            class="modelus"
            type="button"
          >
            %
          </button>
          <button
            class="seven"
            type="button"
          >
            7
          </button>
          <button
            class="eight"
            type="button"
          >
            8
          </button>
          <button
            class="nine"
            type="button"
          >
            9
          </button>
          <button
            class="four"
            type="button"
          >
            4
          </button>
          <button
            class="five"
            type="button"
          >
            5
          </button>
          <button
            class="six"
            type="button"
          >
            6
          </button>
          <button
            class="1"
            type="button"
          >
            1
          </button>
          <button
            class="two"
            type="button"
          >
            2
          </button>
          <button
            class="three"
            type="button"
          >
            3
          </button>
          <button
            class="zero"
            type="button"
          >
            0
          </button>
          <button
            class="dot"
            type="button"
          >
            .
          </button>
        </div>
        <div
          class="right-section"
        >
          <button
            class="plus"
            type="button"
          >
            /
          </button>
          <button
            class="plus"
            type="button"
          >
            x
          </button>
          <button
            class="plus"
            type="button"
          >
            -
          </button>
          <button
            class="plus"
            type="button"
          >
            +
          </button>
          <button
            class="plus"
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
